File: java/src/main/java/global/thalion/ttio/NMRSpectrum.java
/*
 * TTI-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio;

import java.util.Map;

/**
 * 1-D NMR spectrum: chemical-shift + intensity arrays plus nucleus
 * type and spectrometer frequency in MHz.
 *
 * <p><b>API status:</b> Stable.</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIONMRSpectrum}, Python
 * {@code ttio.nmr_spectrum.NMRSpectrum}.</p>
 *
 * @since 0.6
 */
public class NMRSpectrum extends Spectrum {
    private final String nucleusType;
    private final double spectrometerFrequencyMHz;

    public NMRSpectrum(double[] chemicalShiftValues, double[] intensityValues,
                       int indexPosition, double scanTimeSeconds,
                       String nucleusType, double spectrometerFrequencyMHz) {
        super(Map.of(
            "chemical_shift", SignalArray.ofDoubles(chemicalShiftValues),
            "intensity", SignalArray.ofDoubles(intensityValues)
        ), indexPosition, scanTimeSeconds);
        this.nucleusType = nucleusType;
        this.spectrometerFrequencyMHz = spectrometerFrequencyMHz;
    }

    /** Returns the {@code "chemical_shift"} {@link SignalArray}. */
    public SignalArray chemicalShiftArray() { return signalArray("chemical_shift"); }

    /** Returns the {@code "intensity"} {@link SignalArray}. */
    public SignalArray intensityArray() { return signalArray("intensity"); }

    /** Convenience: chemical-shift values as a primitive array. */
    public double[] chemicalShiftValues() { return chemicalShiftArray().asDoubles(); }

    /** Convenience: intensity values as a primitive array. */
    public double[] intensityValues() { return intensityArray().asDoubles(); }

    /** Nucleus type, e.g. {@code "1H"}, {@code "13C"}, {@code "31P"}. */
    public String nucleusType() { return nucleusType; }

    /** Spectrometer frequency in MHz. */
    public double spectrometerFrequencyMHz() { return spectrometerFrequencyMHz; }
}

File: python/src/ttio/nmr_spectrum.py
"""``NMRSpectrum`` — 1-D NMR subclass with chemical-shift + intensity."""
from __future__ import annotations

from dataclasses import dataclass

from .signal_array import SignalArray
from .spectrum import Spectrum


@dataclass(slots=True)
class NMRSpectrum(Spectrum):
    """A 1-D NMR spectrum with ``chemical_shift`` and ``intensity`` channels.

    Parameters
    ----------
    nucleus_type : str, default ""
        Nucleus type (``"1H"``, ``"13C"``, ``"31P"``, ...).
    spectrometer_frequency_mhz : float, default 0.0
        Spectrometer frequency in MHz.
    *plus all base :class:`Spectrum` parameters*

    Notes
    -----
    API status: Stable.

    Cross-language equivalents
    --------------------------
    Objective-C: ``TTIONMRSpectrum`` · Java:
    ``global.thalion.ttio.NMRSpectrum``.
    """

    CHEMICAL_SHIFT = "chemical_shift"
    INTENSITY = "intensity"

    nucleus_type: str = ""
    spectrometer_frequency_mhz: float = 0.0

    @property
    def chemical_shift_array(self) -> SignalArray:
        """Return the ``"chemical_shift"`` :class:`SignalArray`."""
        return self.signal_array(self.CHEMICAL_SHIFT)

    @property
    def intensity_array(self) -> SignalArray:
        """Return the ``"intensity"`` :class:`SignalArray`."""
        return self.signal_array(self.INTENSITY)
